core/kernel/tee_ta_manager.c

```diff
@@ -823,6 +837,137 @@ TEE_Result tee_ta_invoke_command(TEE_ErrorOrigin *err,
 	return res;
 }
 
+#if defined(CFG_TA_STATS)
+static TEE_Result tee_ta_dump_memstats(struct tee_ta_session *s,
+				       uint32_t cancel_req_to,
+				       struct tee_ta_param *param)
+{
+	TEE_Result res = TEE_SUCCESS;
+	struct tee_ta_ctx *ctx = NULL;
+	struct ts_ctx *ts_ctx = NULL;
+
+	ts_ctx = s->ts_sess.ctx;
+	if (ts_ctx)
+		ctx = ts_to_ta_ctx(ts_ctx);
+
+	if (ctx->panicked)
+		return TEE_ERROR_TARGET_DEAD;
+
+	if (tee_ta_try_set_busy_nolock(ctx)) {
+		s->param = param;
+		set_invoke_timeout(s, cancel_req_to);
+		res = ts_ctx->ops->dump_mem_stats(&s->ts_sess);
+		s->param = NULL;
+		tee_ta_clear_busy_nolock(ctx);
+	} else {
+		/* Deadlock avoided */
+		res = TEE_ERROR_BUSY;
+	}
+
+	return res;
+}
+
+TEE_Result tee_ta_instance_stats(void *buff,
+				 uint32_t *buff_size)
+{
+	TEE_Result res = TEE_SUCCESS;
+	struct tee_ta_param param = { };
+	struct tee_ta_dump_stats *dump_stats = NULL;
+	struct tee_ta_session *sess = NULL;
+	struct tee_ta_session_head *open_sessions = NULL;
+	struct tee_ta_ctx *ctx = NULL;
+	uint32_t count = 0;
+	uint32_t size = 0;
+	uint32_t actual_size = 0;
+
+	if (!buff_size)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	mutex_lock(&tee_ta_mutex);
```

Comment:
You're holding this lock all the time while entering each TA. That's a quite big critical section.

Reply:
It does go through the TA instances but only when it's user TA and this user TA has open running session, it calls into TA user mode. It's not the case of calling into each TA user mode.

Regarding of calling to user mode in the critical session, it looks we have to do it, the session is from the global variable -- tee_ctxes/open_sessions and need the context to call into its user mode.
Second, the scenario of this function calling is developer want to dump the TA's runtime context for a debug purpose. The function is not in routine running, but only on demand and triggered by developer.

Reply:
> It does go through the TA instances but only when it's user TA and this user TA has open running session, it calls into TA user mode. It's not the case of calling into each TA user mode.

All TA instances except keep-alive ones are removed when the last session is closed, so it's pretty much all loaded TAs you're entering.

> 
> Regarding of calling to user mode in the critical session, it looks we have to do it, the session is from the global variable -- tee_ctxes/open_sessions and need the context to call into its user mode.

Well, that doesn't prevent us in `tee_ta_invoke_command()` and `tee_ta_open_session()`.

> Second, the scenario of this function calling is developer want to dump the TA's runtime context for a debug purpose. The function is not in routine running, but only on demand and triggered by developer.

That's a slippery slope. The worst things in the code have a tendency to be used as an example of how to do things from time to time.

This can be done without holding the `tee_ta_mutex()` while entering each TA.

Reply:
> it's pretty much all loaded TAs you're entering.
I check `is_user_ta_ctx(&ctx->ts_ctx)` before enter TA...

> Well, that doesn't prevent us in `tee_ta_invoke_command()` and `tee_ta_open_session()`.
it looks prevent `tee_ta_open_session()` which will add a new session to `open_sessions` and the link structure of `open_sessions`.  See below:
`tee_ta_open_session()` --> `tee_ta_init_session()` --> `mutex_lock(&tee_ta_mutex)`;
`tee_ta_invoke_command()` looks fine as there is no `tee_ctxes/open_sessions` access.

> This can be done without holding the `tee_ta_mutex()` while entering each TA.
I know but `ts_ctx->ops->dump_mem_stats(&s->ts_sess)` now is embedded in the `tee_ctxes/open_sessions` iteration. If we unlock `tee_ta_mutex` right before entering user TA and relock it after user TA return, this will give a gap that the session might be removed from `open_sessions`. Do you think this is a valid concern?


I think i might need to further understand your worries about holding `tee_ta_mutex()` while entering user TA in debug mode. Can you explain more?

Reply:
> > Well, that doesn't prevent us in `tee_ta_invoke_command()` and `tee_ta_open_session()`.
> 
> it looks prevent `tee_ta_open_session()` which will add a new session to `open_sessions` and change the link structure of `open_sessions`. See below: `tee_ta_open_session()` --> `tee_ta_init_session()` --> `mutex_lock(&tee_ta_mutex)`; `tee_ta_invoke_command()` looks fine as there is no `tee_ctxes/open_sessions` access.

tee_ta_invoke_command() relies on tee_ta_get_session() to gain exclusive access to the session. You need to do the same here, except that you need a version of that function that works with the mutex already held and also takes a valid session pointer. After that, you can unlock the mutex and do the call into the TA.

> 
> > This can be done without holding the `tee_ta_mutex()` while entering each TA.
> 
> I know but `ts_ctx->ops->dump_mem_stats(&s->ts_sess)` now is embedded in the `tee_ctxes/open_sessions` iteration. If we unlock `tee_ta_mutex` right before entering user TA and relock it after user TA return, this will give a gap that the session might be removed from `open_sessions`. Do you think this is a valid concern?
> 
> I think i might need to further understand your worries about holding `tee_ta_mutex()` while entering user TA in debug mode. Can you explain more?

Critical sections should be minimized, it's good style and also doesn't make people reading the code wonder why such a suboptimal solution was chosen. So the approach you're proposing is, to be blunt, not nice and sets a terrible example for others to follow.

The session handling that we have today isn't perfect, I suspect that tee_ta_get_session() and tee_ta_set_busy() could be combined in a way to do both at the same time without releasing the mutex in between. However, that's not in scope here, unless you feel like taking a stab at that first. ;-)

Reply:
I revisit the code about `tee_ta_invoke_command()` and get to understand how to use `tee_ta_get_session()` and `tee_ta_put_session()`. Thanks.
I worked out a new commit which I record the session id into dump context with holding `tee_ta_mutex()`. Then, call `tee_ta_get_session()` with session id and enter user TA to get heap stats after releasing mutex.